dep0/lib/03_typecheck/src/proof_search.cpp

```diff
@@ -60,51 +85,174 @@ std::size_t search_state_t::hash_t::operator()(expr_t const& x) const
     return hash_code(x);
 }
 
-std::optional<expr_t>
-start_proof_search(
-    env_t const& env,
-    ctx_t const& ctx,
-    expr_t const& type,
+search_task_t::search_task_t(
+    std::size_t const depth,
+    search_state_t& st,
+    std::shared_ptr<expr_t const> target,
     ast::is_mutable_t const is_mutable_allowed,
-    usage_t& usage,
-    ast::qty_t const usage_multiplier)
+    std::shared_ptr<usage_t> usage,
+    ast::qty_t const usage_multiplier,
+    std::function<void(search_task_t&)> task)
+:   m_kind(one_t{std::move(task)}),
+    depth(depth),
+    state(st),
+    env(st.env),
+    ctx(st.ctx),
+    target(std::move(target)),
+    is_mutable_allowed(is_mutable_allowed),
+    usage(std::move(usage)),
+    usage_multiplier(usage_multiplier)
+{ }
+
+bool search_task_t::done() const { return not std::holds_alternative<in_progress_t>(m_status); }
+bool search_task_t::failed() const { return std::holds_alternative<failed_t>(m_status); }
+bool search_task_t::succeeded() const { return std::holds_alternative<succeeded_t>(m_status); }
+
+expr_t& search_task_t::result()
 {
-    search_state_t search_state(is_mutable_allowed);
-    return continue_proof_search(env, ctx, type, search_state, usage, usage_multiplier);
+    auto const succeeded = std::get_if<succeeded_t>(&m_status);
+    assert(succeeded);
+    return succeeded->result;
+}
 
+expr_t const& search_task_t::result() const
+{
+    auto const succeeded = std::get_if<succeeded_t>(&m_status);
+    assert(succeeded);
+    return succeeded->result;
+}
+
+void search_task_t::set_failed()
+{
+    assert(not done());
+    m_status = failed_t{};
+}
+
+void search_task_t::set_result(expr_t x)
+{
+    assert(not done());
+    state.cache.try_emplace(*target, x);
+    m_status = succeeded_t{std::move(x)};
+}
+
+void search_task_t::run()
+{
+    assert(not done());
+    if (done())
+        return;
+    if (depth > 10 or state.expired())
+        return set_failed();
+    match(
+        m_kind,
+        [this] (one_t& one)
+        {
+            one.tactic(*this);
+            if (not done() and std::holds_alternative<one_t>(m_kind))
```

Comment:
Maybe this comment is better instead of the other one 
```suggestion
            // let's check if this has implicitly failed
            if (not done() and std::holds_alternative<one_t>(m_kind))
```